### PR TITLE
Run two parallel simulations of a gz-sim-yarp-plugins-enabled robot

### LINKED ISSUE #147: Try to run two parallel simulations of a gz-sim-yarp-plugins-enabled robot
We can use the existing test infrastructure, and we should use a model without any nws as otherwise launching two different nws with the same ports will fail.

### PR DESCRIPTION
**Pull request by xela-95**
Closes #147

### CODE REVIEW

**Comment by xela-95** [comment]
In https://github.com/robotology/gz-sim-yarp-plugins/pull/153/commits/b460db8597023f5573ad90c78a11f28540919827, two simualtions of the same world are instantiated and launched together, using [TestFixture](https://gazebosim.org/api/sim/8/classgz_1_1sim_1_1TestFixture.html) class provided by `gz-sim`.

**Notes:**
- In this preliminary attempt the model does not contain any `gz-sim-yarp-plugin`;
- [Running the server](https://gazebosim.org/api/sim/8/classgz_1_1sim_1_1Server.html#abf20a9b198865a3d0662c275a80aa526) with the `blocking` parameter set to `false` allow to run each simulation in a separate thread.
- I'm having difficulties to extend the method using a `vector` of `TestFixture`s, because when I try to call the `emplace_back` passing the world path, I get a compilation errors. I'm trying to debug this.

**Comment by xela-95** [comment]
With https://github.com/robotology/gz-sim-yarp-plugins/pull/153/commits/8886369671e3827384dcd95bbd7f6994aa05e89b the unit test `ControlBoardOnMultipleGazeboInstances` uses a coupled pendulum model with a controlboard plugin. 

Now the unit test fails, since there is an error when trying to set two devices in `DeviceRegister`  with the same key.

It is then time to make a modification to the way the keys are generated. 

@traversaro if I recall well you suggested prefixing the current device key with the memory address stored into the pointer in the call to:
```cpp
DeviceRegistry::setDevice(std::string deviceDatabaseKey, yarp::dev::PolyDriver* device2add)
```

Moreover, I think that **all** the plugin singletons key should be updated in the same way, since for example, we should be able to store the data for the same controlboard device belonging to the same model used in two parallel simulations.

Do you foresee any troubles in progressing on this path?

**Comment by xela-95** [comment]
In the last 5 commits I've restructured the data management mechanisms of the plugins that previously made use of the singletons (see https://github.com/robotology/gz-sim-yarp-plugins/issues/147#issuecomment-2049366584). Now the only singleton in the codebase is constituted by the `DeviceRegistry` class.

**Comment by xela-95** [comment]
With https://github.com/robotology/gz-sim-yarp-plugins/pull/153/commits/c345e4cf30d89a92a37a45c3802db81fa877803a I updated the unit test to use a double pendulum, such that the reading of the second joint would present numerical differences pretty quickly.

Here's a video of the simulation of such model:

https://github.com/robotology/gz-sim-yarp-plugins/assets/57228872/7e1bb59d-a7eb-494b-9908-24f3e222299b

Now to make the unit test pass I used the uglier way possible, but also the only one that seems to work at the moment, i.e. using the `ECM` memory address in the device id (along the scoped name of the device and the yarpDeviceName property).

I'm restructuring the code and the unit test in order to make them pass with the new modifications introduced, thing that is taking a bit of time.

**Comment by xela-95** [comment]
There are still three tests failing:
- ControlBoardTorqueControlTest (SEGFAULT)
- ControlBoardPositionDirectControlTest (Subprocess aborted)
- ControlBoardPositionControlTest (SEGFAULT)

**Comment by xela-95** [comment]
What was causing the strange errors (really difficult to debug) the unit test https://github.com/robotology/gz-sim-yarp-plugins/actions/runs/8718567060/job/23916088302#step:12:217 was the use of the `TestHelpers` method https://github.com/robotology/gz-sim-yarp-plugins/blob/9dc1044dc307742d4c5e7ed45d31392931f2440a/tests/test-helpers/TestHelpers.hh#L14-L32

I created this to help retrieve only a vector of devices already casted to the interface needed in the test. For example, if the models has different gz-sim-yarp-plugins defined, I want to perform tests only on `ControlBoardDriver` devices, but the way in which I implemented it does not work and make the tests crash in unexpected ways.

**Comment by xela-95** [comment]
This is the outcome of the [unit test](https://github.com/robotology/gz-sim-yarp-plugins/blob/44229257dfbee80610ec4ceba03e9946b297c122/tests/controlboard/ControlBoardOnMultipleGazeboInstancesTest.cc) in which in one simulation the reference position is 10 degrees, and in the other is 5 degrees:

```
Start 10: ControlBoardOnMultipleGazeboInstancesTest

10: Test command: /home/acroci/repos/gz-sim-yarp-plugins/build/bin/ControlBoardOnMultipleGazeboInstancesTest
10: Working Directory: /home/acroci/repos/gz-sim-yarp-plugins/build/tests/controlboard
10: Environment variables: 
10:  GZ_SIM_SYSTEM_PLUGIN_PATH=/home/acroci/repos/gz-sim-yarp-plugins/build/lib
10: Test timeout computed to be: 1500
10: Running main() from /home/acroci/repos/gz-sim-yarp-plugins/build/_deps/googletest-src/googletest/src/gtest_main.cc
10: [==========] Running 1 test from 1 test suite.
10: [----------] Global test environment set-up.
10: [----------] 1 test from ControlBoardOnMultipleGazeboInstances
10: [ RUN      ] ControlBoardOnMultipleGazeboInstances.StartConcurrentGazeboInstances
10: [WARNING] File specified with a relative path:  ../../../tests/controlboard/conf/gazebo_controlboard_multiple_instances.ini , resolved to:  /home/acroci/repos/gz-sim-yarp-plugins/build/tests/controlboard/../../../tests/controlboard/conf/gazebo_controlboard_multiple_instances.ini . It is recommended to use a URI.
10: [INFO] Yarp configuration file loaded:  /home/acroci/repos/gz-sim-yarp-plugins/build/tests/controlboard/../../../tests/controlboard/conf/gazebo_controlboard_multiple_instances.ini
10: [INFO] gz-sim-yarp-controlboard-system: configuration of device  controlboard_plugin_device2  loaded
10: [DEBUG] gz-sim-yarp-controlboard-system : robot scoped name:  model/triple_pendulum
10: ==================== ecm address for generated id: 0x58490cc5e5e8
10: [DEBUG] gz-sim-yarp-controlboard-system : deviceId:  0x58490cc5e5e8/model/triple_pendulum/controlboard_plugin_device2
10: [DEBUG] gz-sim-yarp-controlboard-system: initialConfiguration:  0.0
10: [DEBUG] |yarp.dev.PolyDriver|gazebo_controlboard| Parameters are (LIMITS (jntPosMax (10000.0)) (jntPosMin (-10000.0))) (POSITION_CONTROL (controlUnits metric_units) (controlLaw joint_pid_gazebo_v1) (kp (3000.0)) (kd (2.0)) (ki (0.100000000000000005551)) (maxInt (99999)) (maxOutput (99999)) (shift (0.0)) (ko (0.0)) (stictionUp (0.0)) (stictionDwn (0.0))) (VELOCITY_CONTROL (controlUnits metric_units) (controlLaw joint_pid_gazebo_v1) (velocityControlImplementationType integrator_and_position_pid) (kp (500.0)) (kd (2.0)) (ki (0.100000000000000005551)) (maxInt (9999)) (maxOutput (9999)) (shift (0.0)) (ko (0.0)) (stictionUp (0.0)) (stictionDwn (0.0))) (controlBoardId "0x58490cc5e5e8/model/triple_pendulum/controlboard_plugin_device2") (device gazebo_controlboard) (initialConfiguration "0.0") (jointNames (joint1)) (yarpDeviceName controlboard_plugin_device2)
10: [INFO] |yarp.dev.PolyDriver|gazebo_controlboard| Created device <gazebo_controlboard>. See C++ class ControlBoardDriver for documentation.
10: [INFO] Found 1 joints from the plugin configuration.
10: [INFO] Found 2 joints from the model description.
10: [INFO] Joint  joint1  added to the control board data.
10: [WARNING] Group TRAJECTORY_GENERATION not found in plugin configuration. Defaults to minimum jerk trajectory.
10: [WARNING] Group TRAJECTORY_GENERATION not found in plugin configuration. Defaults trajectory generation reference speed and accelerations will be applied
10: [DEBUG] Joint  joint1  trajectory generation reference speed:  10  [deg/s]
10: [DEBUG] Joint  joint1  trajectory generation reference acceleration:  10  [deg/s^2]
10: [INFO] Initial configuration found, initializing trajectory generator with it
10: [INFO] Registered YARP device with instance name: 0x58490cc5e5e8/model/triple_pendulum/controlboard_plugin_device2
10: [WARNING] File specified with a relative path:  ../../../tests/controlboard/conf/gazebo_controlboard_multiple_instances.ini , resolved to:  /home/acroci/repos/gz-sim-yarp-plugins/build/tests/controlboard/../../../tests/controlboard/conf/gazebo_controlboard_multiple_instances.ini . It is recommended to use a URI.
10: [INFO] Yarp configuration file loaded:  /home/acroci/repos/gz-sim-yarp-plugins/build/tests/controlboard/../../../tests/controlboard/conf/gazebo_controlboard_multiple_instances.ini
10: [INFO] gz-sim-yarp-controlboard-system: configuration of device  controlboard_plugin_device2  loaded
10: [DEBUG] gz-sim-yarp-controlboard-system : robot scoped name:  model/triple_pendulum
10: ==================== ecm address for generated id: 0x58490e649a88
10: [DEBUG] gz-sim-yarp-controlboard-system : deviceId:  0x58490e649a88/model/triple_pendulum/controlboard_plugin_device2
10: [DEBUG] gz-sim-yarp-controlboard-system: initialConfiguration:  0.0
10: [DEBUG] |yarp.dev.PolyDriver|gazebo_controlboard| Parameters are (LIMITS (jntPosMax (10000.0)) (jntPosMin (-10000.0))) (POSITION_CONTROL (controlUnits metric_units) (controlLaw joint_pid_gazebo_v1) (kp (3000.0)) (kd (2.0)) (ki (0.100000000000000005551)) (maxInt (99999)) (maxOutput (99999)) (shift (0.0)) (ko (0.0)) (stictionUp (0.0)) (stictionDwn (0.0))) (VELOCITY_CONTROL (controlUnits metric_units) (controlLaw joint_pid_gazebo_v1) (velocityControlImplementationType integrator_and_position_pid) (kp (500.0)) (kd (2.0)) (ki (0.100000000000000005551)) (maxInt (9999)) (maxOutput (9999)) (shift (0.0)) (ko (0.0)) (stictionUp (0.0)) (stictionDwn (0.0))) (controlBoardId "0x58490e649a88/model/triple_pendulum/controlboard_plugin_device2") (device gazebo_controlboard) (initialConfiguration "0.0") (jointNames (joint1)) (yarpDeviceName controlboard_plugin_device2)
10: [INFO] |yarp.dev.PolyDriver|gazebo_controlboard| Created device <gazebo_controlboard>. See C++ class ControlBoardDriver for documentation.
10: [INFO] Found 1 joints from the plugin configuration.
10: [INFO] Found 2 joints from the model description.
10: [INFO] Joint  joint1  added to the control board data.
10: [WARNING] Group TRAJECTORY_GENERATION not found in plugin configuration. Defaults to minimum jerk trajectory.
10: [WARNING] Group TRAJECTORY_GENERATION not found in plugin configuration. Defaults trajectory generation reference speed and accelerations will be applied
10: [DEBUG] Joint  joint1  trajectory generation reference speed:  10  [deg/s]
10: [DEBUG] Joint  joint1  trajectory generation reference acceleration:  10  [deg/s^2]
10: [INFO] Initial configuration found, initializing trajectory generator with it
10: [INFO] Registered YARP device with instance name: 0x58490e649a88/model/triple_pendulum/controlboard_plugin_device2
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: [Msg] Found no publishers on /stats, adding root stats topic
10: [Msg] Found no publishers on /stats, adding root stats topic
10: [Msg] Found no publishers on /clock, adding root clock topic
10: [Msg] Found no publishers on /clock, adding root clock topic
10: [Dbg] [SimulationRunner.cc:545] Creating PostUpdate worker threads: 4
10: Creating PostUpdate worker threads: [Dbg] [SimulationRunner.cc:545] Creating PostUpdate worker threads: 4
10: [Dbg] [SimulationRunner.cc:556] Creating postupdate worker thread (0)
10: Creating postupdate worker thread (0)
10: [Dbg] [SimulationRunner.cc:556] Creating postupdate worker thread (1)
10: Creating postupdate worker thread (1)
10: [Dbg] [SimulationRunner.cc:556] Creating postupdate worker thread (2)
10: [Dbg] [SimulationRunner.cc:556] Creating postupdate worker thread (2)
10: [Msg] Serving scene information on [/world/tutorial_controlboard/scene/info]
10: [Msg] Serving scene information on [/world/tutorial_controlboard/scene/info]
10: [Msg] Serving graph information on [/world/tutorial_controlboard/scene/graph]
10: [Msg] Serving graph information on [/world/tutorial_controlboard/scene/graph]
10: [Msg] Serving full state on [/world/tutorial_controlboard/state]
10: [Msg] Serving full state on [/world/tutorial_controlboard/state]
10: [Msg] Serving full state (async) on [/world/tutorial_controlboard/state_async]
10: [Msg] Serving full state (async) on [/world/tutorial_controlboard/state_async]
10: [Msg] Publishing scene information on [/world/tutorial_controlboard/scene/info]
10: [Msg] Publishing scene information on [/world/tutorial_controlboard/scene/info]
10: [Msg] Publishing entity deletions on [/world/tutorial_controlboard/scene/deletion]
10: [Msg] Publishing entity deletions on [/world/tutorial_controlboard/scene/deletion]
10: [Msg] Publishing state changes on [/world/tutorial_controlboard/state]
10: [Msg] Publishing state changes on [/world/tutorial_controlboard/state]
10: [Msg] Publishing pose messages on [/world/tutorial_controlboard/pose/info]
10: [Msg] Publishing pose messages on [/world/tutorial_controlboard/pose/info]
10: [Msg] Publishing dynamic pose messages on [/world/tutorial_controlboard/dynamic_pose/info]
10: [Msg] Publishing dynamic pose messages on [/world/tutorial_controlboard/dynamic_pose/info]
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Waiting for Gazebo simulation to finish...
10: Motion done simulation 1 in 1000 iterations
10: Motion done simulation 2 in 500 iterations
10: Final joint position simulation 1: 9.9983920082
10: Final joint position simulation 2: 4.9983734197
10: [Dbg] [SimulationRunner.cc:572] [Dbg] [SimulationRunner.cc:572] Exiting postupdate worker thread (1)
10: Exiting postupdate worker thread ([Dbg] [SimulationRunner.cc:572] Exiting postupdate worker thread (2)Exiting postupdate worker thread (0
10: )[Dbg] [SimulationRunner.cc:572] Exiting postupdate worker thread (0[Dbg] [SimulationRunner.cc:572] 0))[Dbg] [SimulationRunner.cc:572] Exiting postupdate worker thread (2)
10: 
10: 1)
10: [       OK ] ControlBoardOnMultipleGazeboInstances.StartConcurrentGazeboInstances (12327 ms)
10: [----------] 1 test from ControlBoardOnMultipleGazeboInstances (12327 ms total)
10: 
10: [----------] Global test environment tear-down
10: [==========] 1 test from 1 test suite ran. (12327 ms total)
10: [  PASSED  ] 1 test.
1/1 Test #10: ControlBoardOnMultipleGazeboInstancesTest ...   Passed   12.39 sec
```

**Comment by xela-95** [comment]
I want to iterate a bit on the DeviceRegistry data structure, I will put again this PR into a draft.

**Comment by traversaro** [comment]
The problem was probably some kind of consequences of us breaking the one definition rule (that is the reason why you always include .h header files with declarations or inline definitions and never .cpp with non-inline definitions). Anyhow, if this now works that is great, is this ready for review?

File: plugins/controlboard/src/ControlBoard.cpp [hunk 1]
```diff
@@ -97,27 +102,21 @@ void ControlBoard::Configure(const Entity& _entity,
...
 
     m_pluginParameters.put(yarp::dev::gzyarp::YarpControlBoardScopedName.c_str(),
                            m_robotScopedName.c_str());
 
-    // Insert the pointer in the singleton handler for retrieving it in the yarp driver
-    ControlBoardDataSingleton::getControlBoardHandler()->setControlBoardData(&(m_controlBoardData));
-
     m_pluginParameters.put("device", "gazebo_controlboard");
-    m_pluginParameters.put("controlBoardId", m_deviceScopedName);
+    // m_pluginParameters.put("controlBoardId", m_deviceId);
```
Comment (traversaro) [inline plugins/controlboard/src/ControlBoard.cpp:119]: Can we remove this?

File: plugins/controlboard/src/ControlBoard.cpp [hunk 2]
```diff
@@ -97,27 +102,21 @@ void ControlBoard::Configure(const Entity& _entity,
...
-
-    m_deviceScopedName
-        = m_robotScopedName + "/" + m_pluginParameters.find("yarpDeviceName").asString();
-    yDebug() << "gz-sim-yarp-controlboard-system : device scoped name: " << m_deviceScopedName;
+    yDebug() << "gz-sim-yarp-controlboard-system : yarpDeviceName: " << yarpDeviceName;
 
     m_modelEntity = _entity;
 
-    m_controlBoardData.controlBoardId = m_deviceScopedName;
+    // m_controlBoardData.controlBoardId = m_deviceId;
```
Comment (traversaro) [inline plugins/controlboard/src/ControlBoard.cpp:113]: Can we remove this?

File: plugins/controlboard/include/ControlBoard.hh [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 #include <ControlBoardData.hh>
 
+#include <cstddef>
+#include <gz/msgs/details/wrench.pb.h>
```
Comment (traversaro) [inline plugins/controlboard/include/ControlBoard.hh:6]: Is this required? details tipically indicates private headers that are not meant to be included.

**Review by traversaro — APPROVED** [review]
Minor comments.

**Comment by xela-95** [comment]
> The problem was probably some kind of consequences of us breaking the one definition rule (that is the reason why you always include .h header files with declarations or inline definitions and never .cpp with non-inline definitions). Anyhow, if this now works that is great, is this ready for review?

Yep, I agree with you, I had an unpleasant meeting with "undefined behavior"  in C++ 🤣 

Yep it's ready for review!

File: plugins/controlboard/src/ControlBoard.cpp (hunk 2 above)
Comment (xela-95) [reply plugins/controlboard/src/ControlBoard.cpp]: Solved in https://github.com/robotology/gz-sim-yarp-plugins/pull/153/commits/fd8043fb3c1ec219c31f9d35cd441383ffcfa57b

File: plugins/controlboard/include/ControlBoard.hh (hunk 3 above)
Comment (xela-95) [reply plugins/controlboard/include/ControlBoard.hh]: Solved in Solved in https://github.com/robotology/gz-sim-yarp-plugins/pull/153/commits/fd8043fb3c1ec219c31f9d35cd441383ffcfa57b

File: plugins/controlboard/src/ControlBoard.cpp (hunk 1 above)
Comment (xela-95) [reply plugins/controlboard/src/ControlBoard.cpp]: Solved in https://github.com/robotology/gz-sim-yarp-plugins/pull/153/commits/fd8043fb3c1ec219c31f9d35cd441383ffcfa57b

**Comment by xela-95** [comment]
The [unit test](https://github.com/robotology/gz-sim-yarp-plugins/actions/runs/8752209371/job/24019372386?pr=153#step:11:1309) failing on the CI is quite scary: 
```
ODE INTERNAL ERROR 1: assertion "!colliders_initialized" failed in dInitColliders() [collision_kernel.cpp:168]
```
gz-sim related issue: https://github.com/gazebosim/gz-sim/issues/18

The unit test failing is launching two parallel Gazebo instances of different models (one sphere and one box bouncing on the ground). This error seems related to the ODE collision detector having issues due to some global variable. However it seems to happen only on the apt CI workflow on Ubuntu@Debug.

**Comment by traversaro** [comment]
Ah, that is an old foe. For parallel workflows we should either use another collision detection algorithm when using dart (dart supports several of them if I recall correctly), or using another physics engine. cc @diegoferigo as probably we experienced something similar when working on gym-ignition.

**Comment by xela-95** [comment]
Now the models used by concurrent tests use `dart` both as physics engine and collision detector. All the tests are passing.

Merging 🚀

**Comment by traversaro** [comment]
That is nice for the tests, but it also means that users will start getting the problem if they start using our APIs for parallel computation. We should at least document this once we will document the possibility of running multiserver. See also the discussion in https://github.com/gazebosim/gz-sim/issues/18#issuecomment-2067252636 .

**Pull request merged**